# Laboratorio 1: Exploración y Visualización de Datos

Integrantes: **AGREGUEN SUS NOMBRES COMPLETOS**


## Instrucciones

1.  Trabajen en equipos de dos personas. Salvo excepciones, no se corregirá entregas con menos de dos integrantes.

2.  Modifique este archivo `.ipynb` agregando sus respuestas donde corresponda.

3.  Para cada pregunta **incluya el código fuente que utilizó para llegar a su respuesta. Respuestas sin código no recibirán puntaje.**.

4.  El formato de entrega para esta actividad es un archivo **html**. **Genere un archivo HTML usando Jupyter** y súbalo a U-Cursos. Basta con que un/a integrante haga la entrega. Si ambos/as hacen una entrega en U-Cursos, se revisará cualquiera de éstas.


#### **Se recomienda fuertemente que no usen ChatGPT para resolver la actividad, ya que la experiencia de aprendizaje es mucho mayor si lo hacen por su cuenta.**

## Pongamonos en situación

Eres un **cientifico de datos** que trabaja en **Netflix**.

En los últimos años ha **crecido la cantidad usuarios que ven anime** en la plataforma, por lo que el equipo de contenido de Netflix ha decidido invertir en la producción de series de anime.

Tu jefe te ha pedido que **realices un análisis de los datos de las series de anime** para entender mejor el mercado y poder tomar decisiones informadas.

## Datos

Para realizar esta tarea, te han entregado un dataset con información de animes de la página **MyAnimeList** (MAL).

FUENTE: https://www.kaggle.com/datasets/azathoth42/myanimelist

Tenemos 2 archivos:
- `anime.csv`: Contiene 6 atributos: "title", "type", "source", "episodes", "airing" y "studio", que corresponden a información sobre el anime.
- `anime-stats.csv`: Contiene 4 atributos: "title", "score", "members" (número de usuarios que tienen el anime en su lista) y "favorites", que corresponden a estadísticas sobre la popularidad de cada anime.

In [3]:
import pandas as pd
import matplotlib.pyplot as plt

Para que la siguiente celda funcione, los archivos CSV deben estar en la misma carpeta que este notebook.

In [ ]:
df_anime = pd.read_csv('anime.csv', encoding="UTF-8")
df_stats = pd.read_csv('anime-stats.csv', encoding="UTF-8")

También puedes cargarlos remotamente de un repositorio usando el link a los archivos:


In [5]:
df_anime = pd.read_csv('https://raw.githubusercontent.com/bacchus00/cositas/main/anime.csv', encoding="UTF-8")
df_stats = pd.read_csv('https://raw.githubusercontent.com/bacchus00/cositas/main/anime-stats.csv', encoding="UTF-8")

Cualquiera de estas 2 formas es válida para cargar los datos.

## Exploración básica


Para hacernos una idea de como se ven los datos, lo más usual es mostrar los primeros registros de cada dataset.

1. Muestre los primeros 5 registros de cada dataset.

In [6]:
df_anime.head(5)

title type    source  episodes  airing            studio
0        Inu x Boku SS   TV     Manga        12   False  David Production
1     Seto no Hanayome   TV     Manga        26   False             Gonzo
2   Shugo Chara!! Doki   TV     Manga        51   False         Satelight
3        Princess Tutu   TV  Original        38   False    Hal Film Maker
4  Bakuman. 3rd Season   TV     Manga        25   False         J.C.Staff

In [7]:
df_stats.head(5)

title  score  members  favorites
0        Inu x Boku SS   7.63   283882       2809
1     Seto no Hanayome   7.89   204003       2579
2   Shugo Chara!! Doki   7.55    70127        802
3        Princess Tutu   8.21    93312       3344
4  Bakuman. 3rd Season   8.67   182765       2082

Siempre es útil conocer **cuántos datos y cuántas columnas** tenemos.

Esto es importante para poder responder
- ¿Tenemos suficientes datos?
- ¿Tenemos suficientes columnas?
- ¿Será muy caro computacionalmente trabajar con estos datos?

2. ¿Cuales son las dimensiones de cada dataset? en otras palabras, ¿cuántas filas y columnas tiene cada uno?

In [ ]:
df_anime.shape

# Existen 
# filas: 14478 
# columnas: 6

(14478, 6)

In [ ]:
df_stats.shape
# Existen 
# filas: 14478 
# columnas: 4

(14478, 4)

3. Además de ver algunos registros y las dimensiones, también es útil conocer algunas **estadísticas básicas** de los datos. Usa una función que nos permita **ver las estadísticas básicas** de **todas las columnas**.

In [12]:
df_anime.describe(include='all')

title   type   source      episodes airing          studio
count         14478  14478    14478  14478.000000  14478            8544
unique        14477      7       16           NaN      2             456
top     Hinamatsuri     TV  Unknown           NaN  False  Toei Animation
freq              2   4271     4210           NaN  14127             735
mean            NaN    NaN      NaN     11.308399    NaN             NaN
std             NaN    NaN      NaN     43.443451    NaN             NaN
min             NaN    NaN      NaN      0.000000    NaN             NaN
25%             NaN    NaN      NaN      1.000000    NaN             NaN
50%             NaN    NaN      NaN      1.000000    NaN             NaN
75%             NaN    NaN      NaN     12.000000    NaN             NaN
max             NaN    NaN      NaN   1818.000000    NaN             NaN

In [13]:
df_stats.describe(include='all')

title         score       members      favorites
count         14478  14478.000000  1.447800e+04   14478.000000
unique        14477           NaN           NaN            NaN
top     Hinamatsuri           NaN           NaN            NaN
freq              2           NaN           NaN            NaN
mean            NaN      6.142482  2.296640e+04     311.649606
std             NaN      1.463981  7.498136e+04    2615.554211
min             NaN      0.000000  0.000000e+00       0.000000
25%             NaN      5.550000  2.450000e+02       0.000000
50%             NaN      6.370000  1.679500e+03       2.000000
75%             NaN      7.060000  1.037900e+04      23.000000
max             NaN     10.000000  1.456378e+06  106895.000000

## Posibles errores en los datos

Al construir los datasets, es posible que se hayan generado **datos duplicados**, ya sea porque alguien ingresó dos veces la misma información, o porque se generó un error en la recolección de datos. Es importante **detectar y decidir qué hacer con estos datos**.

4.  ¿Existen filas duplicadas en alguno de los dataframes?


In [ ]:
df_anime.duplicated()

# No existen duplicados en las filas

0        False
1        False
2        False
3        False
4        False
         ...  
14473    False
14474    False
14475    False
14476    False
14477    False
Length: 14478, dtype: bool

In [ ]:
df_stats.duplicated()

# No existen duplicados en las filas

0        False
1        False
2        False
3        False
4        False
         ...  
14473    False
14474    False
14475    False
14476    False
14477    False
Length: 14478, dtype: bool

Otro caso común es que existan **valores nulos** en los datos. Al igual que antes, es importante **identificarlos y decidir qué hacer con ellos**.

5.  ¿Hay filas con valores nulos en los dataframes?

In [ ]:
df_anime.info()

# Existen filas que contienen en la columna 'studio'
# un valor nulo, esto para 5934 filas

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14478 entries, 0 to 14477
Data columns (total 6 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   title     14478 non-null  object
 1   type      14478 non-null  object
 2   source    14478 non-null  object
 3   episodes  14478 non-null  int64 
 4   airing    14478 non-null  bool  
 5   studio    8544 non-null   object
dtypes: bool(1), int64(1), object(4)
memory usage: 579.8+ KB


In [ ]:
df_stats.info()

# No contiene filas con valores nulos

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14478 entries, 0 to 14477
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   title      14478 non-null  object 
 1   score      14478 non-null  float64
 2   members    14478 non-null  int64  
 3   favorites  14478 non-null  int64  
dtypes: float64(1), int64(2), object(1)
memory usage: 452.6+ KB


Consideremos que estamos en la situación en la que **decidimos eliminar los valores nulos**.

6. En caso de existir valores nulos, elimine las filas que los contengan. ¿Cuántas filas quedan en cada dataset luego de eliminar los valores nulos?

In [ ]:
df_anime.dropna()

# Ya vimo que habia filas con nulos, al usar el comando '.dropna()'
# solo quedan 8544

title     type  \
0                                          Inu x Boku SS       TV   
1                                       Seto no Hanayome       TV   
2                                     Shugo Chara!! Doki       TV   
3                                          Princess Tutu       TV   
4                                    Bakuman. 3rd Season       TV   
...                                                  ...      ...   
14461  Tobidashi wa Abunaizo! Mushi Mushi Mura no Kou...      OVA   
14464                                Mr. Men Little Miss      ONA   
14469                                     Ichinen Ikkumi      OVA   
14470  Mushi Mushi Mura no Nakama-tachi: Minna Ii Tok...      OVA   
14477                         Senjou no Valkyria Special  Special   

             source  episodes  airing            studio  
0             Manga        12   False  David Production  
1             Manga        26   False             Gonzo  
2             Manga        51   False         Satelight  
3          Original        38   False    Hal Film Maker  
4             Manga        25   False         J.C.Staff  
...             ...       ...     ...               ...  
14461       Unknown         1   False    Toei Animation  
14464  Picture book         0    True            Sanrio  
14469       Unknown         1   False    Toei Animation  
14470      Original         1   False    Toei Animation  
14477       Unknown         1   False      A-1 Pictures  

[8544 rows x 6 columns]

In [26]:
df_stats.dropna()

# Queda igual

title  score  members  favorites
0                              Inu x Boku SS   7.63   283882       2809
1                           Seto no Hanayome   7.89   204003       2579
2                         Shugo Chara!! Doki   7.55    70127        802
3                              Princess Tutu   8.21    93312       3344
4                        Bakuman. 3rd Season   8.67   182765       2082
...                                      ...    ...      ...        ...
14473             Gutchonpa Omoshiro Hanashi   5.50       76          0
14474                   Geba Geba Shou Time!   4.60       90          0
14475            Godzilla: Hoshi wo Kuu Mono   0.00     1741          0
14476  Nippon Mukashibanashi: Sannen Netarou   6.00       32          0
14477             Senjou no Valkyria Special   5.15      476          0

[14478 rows x 4 columns]

## Juntemos nuestros datos

Con el objetivo de tener toda la información en un solo dataframe, y así poder responder preguntas más complejas, **unamos ambos dataframes**.

7. Junte ambos dataframes en uno solo **a través de la columna "title"** con un **inner join**, y **muestre los primeros 5 registros** de este nuevo dataframe.

In [27]:
pd.merge(df_anime, df_stats, on="title").head(5)

title type    source  episodes  airing            studio  \
0        Inu x Boku SS   TV     Manga        12   False  David Production   
1     Seto no Hanayome   TV     Manga        26   False             Gonzo   
2   Shugo Chara!! Doki   TV     Manga        51   False         Satelight   
3        Princess Tutu   TV  Original        38   False    Hal Film Maker   
4  Bakuman. 3rd Season   TV     Manga        25   False         J.C.Staff   

   score  members  favorites  
0   7.63   283882       2809  
1   7.89   204003       2579  
2   7.55    70127        802  
3   8.21    93312       3344  
4   8.67   182765       2082

## Obtengamos información de los datos

Ahora tenemos que entender un poco más sobre los datos que tenemos, para que el equipo de contenido pueda tomar decisiones informadas.

Para esto, tu jefe te ha solicitado las siguientes tareas:

8. "Necesitamos saber cuales son los animes más populares." Para esto, **entrega los 5 animes con más miembros** (número de usuarios que tienen el anime en su lista).

In [28]:
df_stats.sort_values(by="members", ascending=False).head(5)

title  score  members  favorites
7913                        Death Note   8.67  1456378      88696
8123                Shingeki no Kyojin   8.49  1345842      64032
6296                  Sword Art Online   7.64  1301616      50042
2555  Fullmetal Alchemist: Brotherhood   9.25  1199091     106895
8863                     One Punch Man   8.73  1020754      30747

9. Esta información puede ser muy valiosa para identificar alguna característica en común de los animes más populares. **Realiza una observación de alguna característica en común, o algo relevante, de los 5 animes** obtenidos en la pregunta anterior.

Tienen un Score bastante alto, con el mas bajo de 7.64 y el mas alto de 9.25 con una media de aproximadamente 8.6.

10. Además de un anime popular, el equipo te dice que quiere hacer un anime de mucha calidad, osea, que sea muy bien evaluado, en particular buscan que tenga un score mayor a 9.0. **Entrega los con score mayor a 9.0**.

In [29]:
df_stats[df_stats["score"] > 9.0]

title  score  members  \
418                               Hunter x Hunter (2011)   9.11   720920   
548    Manichi ga Tsurai Kimochi Wakarimasu ka: Yurus...  10.00       29   
1028                                      Kimi no Na wa.   9.19   730076   
1252                                            Gintama°   9.15   198744   
1307                          3-gatsu no Lion 2nd Season   9.10    85744   
2555                    Fullmetal Alchemist: Brotherhood   9.25  1199091   
2783                                Clannad: After Story   9.01   607387   
2999                                         Steins;Gate   9.14  1010330   
3129                                   Okaachan Gomen ne   9.20       73   
5783                            Gintama&#039;: Enchousen   9.07   123572   
6177                                       Steins;Gate 0   9.15   223989   
6553                                             Gintama   9.01   459031   
6572                                      Koe no Katachi   9.04   511915   
6615                          Owarimonogatari 2nd Season   9.02   113928   
8992                                       Gintama&#039;   9.11   215527   
10879                                           Gintama.   9.02    91695   
12724                               Ginga Eiyuu Densetsu   9.11   124993   
13230  Gintama Movie 2: Kanketsu-hen - Yorozuya yo Ei...   9.04   105458   
14208                                       Hashire John   9.50       41   
14317  Kuunyan no Koutsuu Anzen: Tadashii Jitensha no...  10.00       34   
14403                            Dokidoki Little Ooyasan   9.52      357   
14407                                         Ittekimasu  10.00       30   
14454           Tat-chan - Momo-chan no Fushigina Taiken   9.50       35   
14467               Kyouiku Eiga-sai Jushou Anime Series  10.00       30   

       favorites  
418        64375  
548            0  
1028       34912  
1252        5668  
1307        2257  
2555      106895  
2783       47380  
2999       92423  
3129           0  
5783        1509  
6177        1882  
6553       33781  
6572       19815  
6615        2432  
8992        4549  
10879       1237  
12724       8567  
13230       1395  
14208          0  
14317          0  
14403         21  
14407          0  
14454          0  
14467          0

11. Ahora tu equipo desea analizar a la competencia. Para esto **entrega los 5 estudios de anime con mejor score promedio entre sus animes**.

12. Después de que te preguntaran por los animes que tienen mejor score, se te ocurre crear una columna que indique cuales animes son "buenos". Para esto, el área de producto te indica que un "mal" anime tiene score bajo 4.0, uno "intermedio" tiene score entre 4.0 y 8.0, y uno "bueno" tiene score sobre 8.0. **Usa el método apply para crear una nueva columna "quality" que tenga como valor "Malo", "Intermedio" o "Bueno" según la definición anterior. Luego, muestra los 5 primeros registros**

13. Rodrigo, un miembro del equipo, te dice: "Si deseamos que la gente nos tenga en sus favoritos, ¿qué tipo de anime deberíamos hacer? Para cada tipo de anime (TV, OVA, Movie, etc) podríamos sumar la cantidad de favoritos, y hacer el tipo de anime que tenga la suma más grande". Sin embargo, tú te percatas de que **hay un problema con lo que él propone**. ¿Cuál es este problema?

14. Para mostrarle a Rodrigo el problema con la solución que él propuso, **entrega la cantidad de animes por cada tipo**.

15. Tu jefe te dice que tuvo una visión en un sueño, te dice que sabe como hacer el anime más famoso. Quiere hacer un anime de tipo OVA, donde la fuente (columna source) sea Novel. **Entrega las columnas title, members, favorites, y quality de todos los animes que cumplan con estas características**.

16. Para verificar si tu jefe tiene sueños extraños, o si realmente tiene una visión, **calcula y muestra el promedio de los miembros, el promedio de los favoritos, y la categoría de quality más repetida** de los animes obtenidos en la pregunta anterior.

# Visualicemos los datos

17. Muestre el histograma de todas las puntuaciones de animes

18. Muestre la densidad de puntajes para los principales tipos de anime (TV, OVA, Movie)

19. ¿Cuál tipo de anime es más exitoso? ¿Se está tomando en cuenta la cantidad de animes por tipo al afirmar eso?

R:

Y eso es todo por hoy :)


![](https://i.imgur.com/ghKLs78.jpeg)